src/Reflection/InitializerExprTypeResolver.php

```diff
@@ -374,34 +383,33 @@ public function getConcatType(Expr $left, Expr $right, callable $getTypeCallback
 			return $leftStringType->append($rightStringType);
 		}
 
+		$leftConstantStrings = TypeUtils::getConstantStrings($leftStringType);
+		$rightConstantStrings = TypeUtils::getConstantStrings($rightStringType);
+		$combinedConstantStringsCount = count($leftConstantStrings) * count($rightConstantStrings);
+
 		// we limit the number of union-types for performance reasons
-		if ($leftStringType instanceof UnionType && count($leftStringType->getTypes()) <= 16 && $rightStringType instanceof ConstantStringType) {
-			$constantStrings = TypeUtils::getConstantStrings($leftStringType);
-			if (count($constantStrings) > 0) {
-				$strings = [];
-				foreach ($constantStrings as $constantString) {
-					if ($constantString->getValue() === '') {
-						$strings[] = $rightStringType;
+		if ($combinedConstantStringsCount > 0 && $combinedConstantStringsCount <= 16) {
```

Comment:
I noticed that there is also `CALCULATE_SCALARS_LIMIT` in this class, but it was not used here, so I kept the previous limit.